eppo_core/src/eval/eval_assignment.rs

```diff
@@ -214,84 +205,32 @@ impl Flag {
-            sa.entry("id".into()).or_insert_with(|| subject_key.into());
-            sa
-        };
+        let subject = Subject::new(subject_key.clone(), subject_attributes.clone());
 
-        let Some((allocation, split)) = self.allocations.iter().find_map(|allocation| {
+        let Some(split) = self.allocations.iter().find_map(|allocation| {
             let mut visitor = visitor.visit_allocation(allocation);
-            let result = allocation.get_matching_split(
-                &mut visitor,
-                subject_key,
-                &subject_attributes_with_id,
-                self.total_shards,
-                now,
-            );
+            let result = allocation.get_matching_split(&mut visitor, &subject, now);
             visitor.on_result(result);
-            result.ok().map(|split| (allocation, split))
+            result.ok()
         }) else {
             return Err(EvaluationFailure::DefaultAllocationNull);
         };
 
-        let variation = self.variations.get(&split.variation_key).ok_or_else(|| {
-            log::warn!(target: "eppo",
-                       flag_key:display = self.key,
-                       subject_key,
-                       variation_key:display = split.variation_key;
-                       "internal: unable to find variation");
-            EvaluationFailure::Error(EvaluationError::UnexpectedConfigurationError)
-        })?;
-
-        visitor.on_variation(variation);
-
-        let assignment_value = variation
-            .value
-            .to_assignment_value(self.variation_type)
-            .ok_or_else(|| {
-                log::warn!(target: "eppo",
-                           flag_key:display = self.key,
-                           subject_key,
-                           variation_key:display = split.variation_key;
-                           "internal: unable to convert Value to AssignmentValue");
-                EvaluationFailure::Error(EvaluationError::UnexpectedConfigurationError)
-            })?;
-
-        let event = allocation.do_log.then(|| AssignmentEvent {
-            feature_flag: self.key.clone(),
-            allocation: allocation.key.clone(),
-            experiment: format!("{}-{}", self.key, allocation.key),
-            variation: variation.key.clone(),
-            subject: subject_key.to_owned(),
-            subject_attributes: subject_attributes.clone(),
-            timestamp: now.to_rfc3339(),
-            meta_data: [
-                ("sdkName".to_owned(), meta.name.to_owned()),
-                ("sdkVersion".to_owned(), meta.version.to_owned()),
-                (
-                    "eppoCoreVersion".to_owned(),
-                    env!("CARGO_PKG_VERSION").to_owned(),
-                ),
-            ]
-            .into(),
-            extra_logging: split.extra_logging.clone(),
-            evaluation_details: None,
-        });
+        let (value, event_base) = split.result.clone()?;
```

Comment:
📓 fetch pre-computed assignment event

Reply:
The `event_base` is precomputed and contains the following fields:
```rust
struct AssignmentEventBase {
    feature_flag: Str, // <- flag key
    allocation: Str, // <- allocation key
    experiment: String, // <- {flag}-{allocation}
    variation: Str, // variation key from split
    meta_data: EventMetaData, // constant {sdk_name,sdk_version,core_version}
    extra_logging: HashMap<String, String>, // extra logging from split
}
```

None of these values are dependent on subject or specific evaluation call, and the split determines all the fields here. Assignment value is also dictated by the selected split.

More abstractly, we pre-compute a map of {split -> evaluation result} (technically, a decision tree), and evaluation is just a matter of selecting the split and returning the associated result.

Reply:
@rasendubi If the assignment base is already pre-computed what benefit is there to `cloning` it and making another copy, if the goal of this optimization is to reduce the memory allocated during evaluation? 

Is it because of Rust's ownership model? Can we just reference the value and create an `AssignmentEvent` by reference?

Reply:
We're not making another copy because it's behind an `Arc` (atomic reference counter). When we're cloning the `Arc`, it just increases the reference counter without copying the data.

I believe it's similar to Swift, except it's just a normal/explicit type in Rust (it's not built-in)